src/engine/ContextDependentPiecewiseLinearConstraint.h

```diff
@@ -0,0 +1,246 @@
+ **  * _cdConstraintActive
+ **  * _cdPhaseStatus
+ **  * _cdInfeasibleCases
+ **
+ ** PhaseStatus enumeration is used to keep track of current phase of the
+ ** object, as well as to record previously explored phases. This replaces
+ ** moving PiecewiseLinearCaseSplits around.
+ **
+ ** ContextDependentPiecewiseLinearConstraint stores locally which phases have
+ ** been explored so far using _cdInfeasibleCases. To communicate with the
+ ** search it implements two methods:
+ **   * markInfeasible( PhaseStatus caseId ) - marks explored phase
+ **   * nextFeasibleCase() - obtains next unexplored case if one exists
+ **
+ ** These methods rely on subclass properly handling:
+ **  * _numCases field, which denotes the number of piecewise linear
+ **     segments the constraint has, initialized by the constructor.
+ **  * getAllCases() virtual method returns all possible phases represented using
+ **    PhaseStatus enumeration. The order of returned cases will determine the
+ **    search order, so this method should implement any search heuristics.
+ **/
+
+#ifndef __ContextDependentPiecewiseLinearConstraint_h__
+#define __ContextDependentPiecewiseLinearConstraint_h__
+
+#include "FloatUtils.h"
+#include "ITableau.h"
+#include "List.h"
+#include "Map.h"
+#include "MarabouError.h"
+#include "PiecewiseLinearCaseSplit.h"
+#include "PiecewiseLinearConstraint.h"
+#include "PiecewiseLinearFunctionType.h"
+#include "Queue.h"
+#include "Tightening.h"
+#include "context/cdlist.h"
+#include "context/cdo.h"
+#include "context/context.h"
+
+class Equation;
+class BoundManager;
+class ITableau;
+class InputQuery;
+class String;
+
+class ContextDependentPiecewiseLinearConstraint
+   : public virtual PiecewiseLinearConstraint
+{
+public:
+    ContextDependentPiecewiseLinearConstraint();
+    ContextDependentPiecewiseLinearConstraint( unsigned numCases );
+
+    virtual ~ContextDependentPiecewiseLinearConstraint()
+    {
+        cdoCleanup();
+    }
+
+    /*
+      Turn the constraint on/off.
+    */
+    void setActiveConstraint( bool active );
+    bool isActive() const;
+
+    /*
+       Returns a list of all cases of this constraint and is used by the
+       nextFeasibleCase during search. The order of returned cases affects the
+       search, and this method is where related heuristics should be
+       implemented.
+     */
+    virtual List<PhaseStatus> getAllCases() const = 0;
+
+    /*
+     * Returns case split corresponding to the given phase/id
+       TODO: Update the signature in PiecewiseLinearConstraint, once the new
+       search is integrated.
+     */
+    virtual PiecewiseLinearCaseSplit getCaseSplit( PhaseStatus caseId ) const = 0;
+
+
+    /*
+      Check if the constraint's phase has been fixed.
+    */
+    virtual bool phaseFixed() const = 0;
+
+    /*
+      If the constraint's phase has been fixed, get the (valid) case split.
+    */
+    virtual PiecewiseLinearCaseSplit getImpliedCaseSplit() const = 0;
```

Comment:
As a side note, we said we wanted to change this name, right? Maybe getImpliedCase, to be consistent with getAllCases()?

Reply:
So I would keep the name as is because there is a distinction in the return type - get*Case(s) all return PhaseStatus enumeration in some form, while this one actually returns an actual PiecewiseLinearCaseSplit. In the long run this should be the only method to return a PiecewiseLinearCaseSplit.